pkg/cli/image/strategy/onerror.go

```diff
@@ -0,0 +1,143 @@
+// sources only after getting an error from the original requested.
+func NewICSPOnErrorStrategy(file string) registryclient.AlternateBlobSourceStrategy {
+	return &onErrorStrategy{
+		icspFile:              file,
+		alternates:            make(map[reference.DockerImageReference][]reference.DockerImageReference),
+		readICSPsFromFileFunc: readICSPsFromFile,
+	}
+}
+
+func (s *onErrorStrategy) FirstRequest(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	return nil, nil
+}
+
+func (s *onErrorStrategy) OnFailure(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if alternates, ok := s.alternates[locator]; ok {
+		return alternates, nil
+	}
+	alternates, err := s.resolve(ctx, locator)
+	if err != nil {
+		return nil, err
+	}
+	if len(alternates) == 0 {
+		return nil, fmt.Errorf("no alternative image references found for image: %s", locator.String())
+	}
+	s.alternates[locator] = alternates
+	return s.alternates[locator], nil
+}
+
+// resolve gathers possible image sources for a given image
+// gathered from ImageContentSourcePolicy file.
+// Image reference of user-given image may be different from original in case of mirrored images.
+func (s *onErrorStrategy) resolve(ctx context.Context, imageRef reference.DockerImageReference) ([]reference.DockerImageReference, error) {
+	if len(s.icspFile) == 0 {
+		return nil, fmt.Errorf("no ImageContentSourceFile specified")
+	}
+	klog.V(5).Infof("Reading ICSP from file %s", s.icspFile)
+	icspList, err := s.readICSPsFromFileFunc(s.icspFile)
+	if err != nil {
+		return nil, err
+	}
+	imageRefList, err := alternativeImageSources(imageRef, icspList)
+	if err != nil {
+		return nil, err
+	}
+	// always add the original as the first reference
+	imageRefList = append([]reference.DockerImageReference{imageRef.AsRepository()}, imageRefList...)
+	return imageRefList, nil
+}
+
+// alternativeImageSources returns unique list of DockerImageReference objects from list of ImageContentSourcePolicy objects
+func alternativeImageSources(imageRef reference.DockerImageReference, icspList []operatorv1alpha1.ImageContentSourcePolicy) ([]reference.DockerImageReference, error) {
+	var imageSources []reference.DockerImageReference
+	klog.V(5).Infof("%v ImageReference added to potential ImageSourcePrefixes from ImageContentSourcePolicy", imageRef.AsRepository())
+	for _, icsp := range icspList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		for _, rdm := range repoDigestMirrors {
+			var err error
+			rdmSourceRef, err := reference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			// AsV2 in the right call is required to ensure we transform docker registry
+			// from docker.io to registry-1.docker.io
+			// TODO: support partial match, iow. quay.io/foo should match quay.io/foo/bar
```

Comment:
Hrm, that wouldn't be correct.  quay.io/foo is simply incorrect as quay.io/foo is not a repository.  Remove this TODO.